crates/compiler/test_gen/src/gen_num.rs

```diff
@@ -2997,161 +2997,199 @@ fn u8_mul_greater_than_i8() {
+        std::f32::NEG_INFINITY,
+        f32
+    );
+    assert_evals_to!(
+        "Num.subSaturated Num.maxF64 -Num.maxF64",
+        std::f64::INFINITY,
+        f64
+    );
+    assert_evals_to!(
+        "Num.subSaturated Num.minF64 -Num.minF64",
+        std::f64::NEG_INFINITY,
+        f64
+    );
+
+    assert_evals_to!(
+        "Num.subSaturated 170_141_183_460_469_231_731dec -1",
+        RocDec::from_str("170141183460469231731.687303715884105727").unwrap(),
+        RocDec
+    );
+    assert_evals_to!(
+        "Num.subSaturated -170_141_183_460_469_231_731dec 1",
+        RocDec::from_str("-170141183460469231731.687303715884105728").unwrap(),
+        RocDec
     );
 }
 
 #[test]
 #[cfg(any(feature = "gen-llvm", feature = "gen-wasm", feature = "gen-dev"))]
 fn mul_saturated() {
+    assert_evals_to!("Num.mulSaturated 200u8 2", 255u8, u8);
+    assert_evals_to!("Num.mulSaturated 100i8 2", 127i8, i8);
+    assert_evals_to!("Num.mulSaturated -100i8 2", -128i8, i8);
+    assert_evals_to!("Num.mulSaturated 40000u16 2", 65535u16, u16);
+    assert_evals_to!("Num.mulSaturated 20000i16 2", 32767i16, i16);
+    assert_evals_to!("Num.mulSaturated -20000i16 2", -32768i16, i16);
+    assert_evals_to!("Num.mulSaturated 3000000000u32 2", 4294967295u32, u32);
+    assert_evals_to!("Num.mulSaturated 2000000000i32 2", 2147483647i32, i32);
+    assert_evals_to!("Num.mulSaturated -2000000000i32 2", -2147483648i32, i32);
+    assert_evals_to!(
+        "Num.mulSaturated 10000000000000000000u64 2",
+        18446744073709551615u64,
+        u64
+    );
     assert_evals_to!(
-        indoc!(
-            r"
-            x : U8
-            x = 20
-            y : U8
-            y = 20
-            Num.mulSaturated x y
-            "
-        ),
-        255,
-        u8
+        "Num.mulSaturated 5000000000000000000i64 2",
+        9223372036854775807i64,
+        i64
     );
     assert_evals_to!(
-        indoc!(
-            r"
-            x : I8
-            x = -20
-            y : I8
-            y = -20
-            Num.mulSaturated x y
-            "
-        ),
-        127,
-        i8
+        "Num.mulSaturated -5000000000000000000i64 2",
+        -9223372036854775808i64,
+        i64
     );
     assert_evals_to!(
-        indoc!(
-            r"
-            x : I8
-            x = 20
-            y : I8
-            y = -20
-            Num.mulSaturated x y
-            "
-        ),
-        -128,
-        i8
+        "Num.mulSaturated -5000000000000000000i64 2",
+        -9223372036854775808i64,
+        i64
     );
+    assert_evals_to!("Num.mulSaturated Num.maxF32 2", std::f32::INFINITY, f32);
+    assert_evals_to!("Num.mulSaturated Num.minF32 2", std::f32::NEG_INFINITY, f32);
+    assert_evals_to!("Num.mulSaturated Num.maxF64 2", std::f64::INFINITY, f64);
+    assert_evals_to!("Num.mulSaturated Num.minF64 2", std::f64::NEG_INFINITY, f64);
+
+    // TODO: This doesn't work anywhere? It returns -1.374607431768211456 : Dec ?
```

Comment:
`Num.mulSaturated` doesn't work as I'd expect with `Dec`, I left a note here, but could use some direction whether this is a bug or not.

Reply:
@bhansconnect would probably be the best person to answer this.

Reply:
if I enable this test code, the following failure is reported:
```
---- gen_num::mul_saturated stdout ----
thread 'gen_num::mul_saturated' panicked at crates/compiler/test_gen/src/helpers/dev.rs:358:13:
assertion `left == right` failed: output is different
  left: RocDec([0, 0, 216, 227, 210, 105, 236, 236, 255, 255, 255, 255, 255, 255, 255, 255], "-1.374607431768211456")
 right: RocDec([255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 127], "170141183460469231731.687303715884105727")
 ```

Reply:
That result is wrapping multiplication

Reply:
ah, I think it's a bug in the bitcode impl then, https://github.com/roc-lang/roc/blob/bfbc9ceef054fbf5ae07e9df530d228936609d42/crates/compiler/builtins/bitcode/src/dec.zig#L431-L434 is just returning `answer.value`. 

Compared to `addSaturated` which checks and returns `max` or `min`  https://github.com/roc-lang/roc/blob/bfbc9ceef054fbf5ae07e9df530d228936609d42/crates/compiler/builtins/bitcode/src/dec.zig#L267-L279

I can make a fix for this in this PR

Reply:
hmm, on second thought, it seems `mulWithOverflow` (which is used for `Num.mulChecked` and `Num.mul`) is incorrect for `RocDec`
```
» Num.mulChecked 170_141_183_460_469_231_731dec 2

Ok -1.374607431768211456 : Result Dec [Overflow]
```

I don't know how to fix this. Is it okay to make a ticket for this, and leave the current PR as is?

Reply:
Yes, please make an Issue and link it. Let me know if you're ready for a review on this.

Reply:
Okay, I made #7276 This change is ready to review.